### DISCUSSION TITLE
data rows from Postgresql Database seem to not map to Pydantic schema

### DISCUSSION THREAD

**Discussion by houmie** [fastapi/fastapi#8874] [Questions] [answered]
I have recently switched to Postgresql from MySQL and have come across an anomaly.

In MySQL I could fetch this from Database and it would be cast into Pydantic schema automatically.

```
async def get_device(device_id: str) -> Device:
    query = DeviceTable.select().where(DeviceTable.c.id == device_id)
    return await db.fetch_one(query)
```

```
class Device(BaseModel):
    id: str
    type: DeviceType
    created_at: datetime
    last_login_at: datetime
    is_banned: bool
```

However when using Postgres this conversion doesn't seem to happen.

For `device.is_banned` I get the error message ```{AttributeError}'Record' object has no attribute 'is_banned'```.

The device object is in this case a Record object ```<databases.backends.postgres.Record object at 0x10d2c9860>``` instead of Pydantic Schema.

Why would FastAPI's behaviour change only because a different database is used? I can't find anything in the documentation about this.

**UPDATE**
I seem to have found a workaround. I have to cast the record explicitly into the Pydantic schema.

```
async def get_device(device_id: str) -> Device:
    query = DeviceTable.select().where(DeviceTable.c.id == device_id)
    record = await db.fetch_one(query)
    return Device(**record)
```

This is quite odd that it doesn't happen automatically. I have to change code because the underlying database has changed. The worse part is that when no record has been found, the cast will crash. Hence I have to make more repetitive changes:

```
async def get_device(device_id: str) -> Optional[Device]:
    query = DeviceTable.select().where(DeviceTable.c.id == device_id)
    record = await db.fetch_one(query)
    if record:
        return Device(**record)
    else:
        return None
```

Any advise please?
Many Thanks,
Houman

**Comment by Mause** [comment]
Hi @houmie - I'm not sure why you think fastapi does automatic casting within your code, it definitely doesn't. The only situation it does "casting" is when you return a model from an endpoint with a schema defined

**Comment by houmie** [comment, marked as answer]
Hi Mause,

Thank you for your reply. 

This is caused by the underlying postgres library being used. I'm guessing there is some magic going on in the Record object of asyncpg. My guess is that the returned Record from the library does not expose the properties. Thus it requires to manually select the attributes to be fed into the pydantic model.

The function I provided above does the casting automatically when aiomysql is used (MySQL is the underlying database). The object I receive from that function is truly a Pydantic "Device" object. 

This has been my findings since I asked the question. :-)
